Created privacy-design.md

johnpallett opened the pull request:
This is a privacy design for the WebXR Device API, designed to inform text for the specification. It draws upon work in the Privacy and Security repo.

Please see the Background section for an explanation of threat vectors and mitigations that informed this design.

A few notes about this revision:
* At time of writing, this revision matches the currently published spec at https://immersive-web.github.io/webxr ... it does not incorporate issues still in discussion.
* It is assumed that inline and immersive sessions can be fulfilled by the same privacy requirements. For example, the 'focused' and 'visible' requirements for viewer pose data are the same for both immersive and inline sessions, though the implications for user experience and privacy are different.
* User consent - when required - is designed as part of session creation. This is done to avoid user fatigue or confusion of having multiple consent prompts during a session, and to address UX concerns surrounding consent prompts after don flows.
* There are several unaddressed threats listed at the end of the Background section.

NellWaliczek commented:
Hey John, thanks so much for posting this to get the ball rolling.  You've covered a lot of really critical information and have been very thorough in laying out the details.  Seriously, thank you for taking the time to do this.

At first glance, I do have a few concerns with several of them statements running the risk of preventing different browsers from making different choices where appropriate for their business needs.  Though, I definitely understand there's a balance to be struck so that we're creating a consistent API and protecting end-users.  Over the next few days I'll give this a deeper look and be sure to get you specific feedback in a timely manner.

Thanks, again!

johnpallett commented:
Thanks Nell! Eager to hear your feedback and looking forward to the
discussion. Please let me know if you have questions in the meantime.

*From: *Nell Waliczek <notifications@github.com>
*Date: *Mon, May 13, 2019 at 5:24 PM
*To: *immersive-web/webxr
*Cc: *johnpallett, Author

Hey John, thanks so much for posting this to get the ball rolling. You've
> covered a lot of really critical information and have been very thorough in
> laying out the details. Seriously, thank you for taking the time to do this.
>
> At first glance, I do have a few concerns with several of them statements
> running the risk of preventing different browsers from making different
> choices where appropriate for their business needs. Though, I definitely
> understand there's a balance to be struck so that we're creating a
> consistent API and protecting end-users. Over the next few days I'll give
> this a deeper look and be sure to get you specific feedback in a timely
> manner.
>
> Thanks, again!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/immersive-web/webxr/pull/638?email_source=notifications&email_token=AF2I7VHLDQEPZEKSCAQBLHDPVIBCXA5CNFSM4HMQ4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVJ5HAA#issuecomment-492032896>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF2I7VBY6S263CEDAMIQTULPVIBCXANCNFSM4HMQ4QCA>
> .
>


-- 
Thanks,
John

asajeffrey commented:
I have a question about user consent, in particular "User consent should only be considered valid for the browsing context within which it was obtained."

Imagine a scenario where a user navigates to a document, provides consent for entering immersive mode, and bookmarks the document, then later visits the bookmark. Do we want to require the browser to ask for consent again whenever the browsing context changed? So user consent needs to be asked for only when the resources for the document have been reclaimed, e.g. due to memory pressure or a process being closed?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   If the reference space is of type [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded), [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) or [eye-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-eye-level), the [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose) data must be limited to ...
+*   If [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose) data is generated by sensors, at least one of the following conditions must be met; the user agent may choose the approach that provides the best user experience:
+    *   User consent must have been obtained, _or_
+    *   Any [XRView](https://immersive-web.github.io/webxr/#xrview) transform data generated by sensors must be [quantized](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#rounding-quantization-or-fuzzing) to prevent sensor fingerprinting, _or_
```
I suspect that Apple (should they choose to implement WebXR) may have concerns with this approach given their [comments](https://github.com/w3c/deviceorientation/issues/57#issuecomment-447429061) during the recent accelerometer data issue.

NellWaliczek replied:
Are there specific implementers who want to do quantization instead of permissions?  If not, I'm hesitant to contort the spec text to provide the option.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+    *   User consent must have been obtained, _or_
+    *   Any [XRView](https://immersive-web.github.io/webxr/#xrview) transform data generated by sensors must be [quantized](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#rounding-quantization-or-fuzzing) to prevent sensor fingerprinting, _or_
+    *   The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.
+*   User consent must have been obtained on devices with configurable interpupillary distance before creating an [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) that will return an [XRView](https://immersive-web.github.io/webxr/#xrview) for each eye.
```
Is this different from what was listed above in the XRSession section?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+    *   Any [XRView](https://immersive-web.github.io/webxr/#xrview) transform data generated by sensors must be [quantized](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#rounding-quantization-or-fuzzing) to prevent sensor fingerprinting, _or_
+    *   The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.
+*   User consent must have been obtained on devices with configurable interpupillary distance before creating an [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) that will return an [XRView](https://immersive-web.github.io/webxr/#xrview) for each eye.
+*   If the device supports configurable or factory-calibrated interpupillary distance that may vary from device to device, then the [XRView](https://immersive-web.github.io/webxr/#xrview) transform data must be rounded to prevent fingerprinting.
```
Rounded to what precision?

NellWaliczek replied:
I don't have any data on this, personally.  Let's add something that explicitly states the precision is at the UAs discretion.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+    *   Any [XRView](https://immersive-web.github.io/webxr/#xrview) transform data generated by sensors must be [quantized](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#rounding-quantization-or-fuzzing) to prevent sensor fingerprinting, _or_
+    *   The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.
+*   User consent must have been obtained on devices with configurable interpupillary distance before creating an [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) that will return an [XRView](https://immersive-web.github.io/webxr/#xrview) for each eye.
+*   If the device supports configurable or factory-calibrated interpupillary distance that may vary from device to device, then the [XRView](https://immersive-web.github.io/webxr/#xrview) transform data must be rounded to prevent fingerprinting.
```
Even if user consent has been granted?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+    *   The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.
+*   User consent must have been obtained on devices with configurable interpupillary distance before creating an [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) that will return an [XRView](https://immersive-web.github.io/webxr/#xrview) for each eye.
+*   If the device supports configurable or factory-calibrated interpupillary distance that may vary from device to device, then the [XRView](https://immersive-web.github.io/webxr/#xrview) transform data must be rounded to prevent fingerprinting.
+*   If the [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) will include multiple [XRView](https://immersive-web.github.io/webxr/#xrview)s for displays whose positions and orientations have been configured by the user (e.g. in a CAVE) then the [XRView](https://immersive-web.github.io/webxr/#xrview) transform data must be rounded to prevent fingerprinting.
```
Rounded to what precision?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   If the device supports configurable or factory-calibrated interpupillary distance that may vary from device to device, then the [XRView](https://immersive-web.github.io/webxr/#xrview) transform data must be rounded to prevent fingerprinting.
+*   If the [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) will include multiple [XRView](https://immersive-web.github.io/webxr/#xrview)s for displays whose positions and orientations have been configured by the user (e.g. in a CAVE) then the [XRView](https://immersive-web.github.io/webxr/#xrview) transform data must be rounded to prevent fingerprinting.
+
+#### Providing XRPose data
```
Is this whole section the same as the one above (minus references to IPD related stuff?)

NellWaliczek replied:
It's a nit, but I think it would help me be sure I wasn't missing something if we actually consolidated the two sections as much as possible.  Then any additional or different requirements can be called out explicitly.  wdyt?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+    *   The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.
+
+#### Creating XRReferenceSpace
+The user agent must verify that all mandatory conditions are satisfied to ensure it can create an [XRReferenceSpace](https://immersive-web.github.io/webxr/#xrreferencespace-interface) and expose [XRReferenceSpace](https://immersive-web.github.io/webxr/#xrreferencespace-interface) data within a given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).
```
>data within a given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).

I'm not sure what your intention is with this portion of the statement since reference spaces are session-bound?

NellWaliczek replied:
Yeah, is that in alignment with your intention?

NellWaliczek replied:
Related to my question https://github.com/immersive-web/webxr/pull/638#discussion_r283964027

johnpallett replied:
Upon review, the use of 'active document' here **is** important to the 2nd mandatory condition: _Currently focused area belongs to a document whose origin is same origin-domain with the origin of the given active document._

The purpose of this requirement is to ensure that Pose or ViewerPose data is only exposed if the focused document is same-origin with the Pose/ViewerPose session. This is a running, real-time requirement to prevent input sniffing.

I'm not sure how to specify this requirement if we remove the concept of an active document, any ideas?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   User consent must have been obtained before creating a [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) or [unbounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-unbounded) reference space. 
+*   User consent is required before creating a [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) reference space type, if the floor level will reflect the real-world location of the user’s floor.
+*   When creating a [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) or [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) reference space type, if the origin of the reference space reflects the real-world location of the user’s floor, or if the floor level is emulated and set by the user to a non-default value, then the co ...
+*   When creating a [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) reference space the [boundsGeometry](https://immersive-web.github.io/webxr/#dom-xrboundedreferencespace-boundsgeometry) must be rounded sufficiently to prevent fingerprinting the user’s bounds. The region represented by the rounded bounds geometry must be a subset of the original bounds. Roundin ...
```
>Rounding to the nearest 5cm is suggested.

Suggested?  Or the minimum required?  Either way, I'm not certain this is really enough to mitigate the concern on systems like Windows MR which allow the user to walk the boundary to define it.

NellWaliczek replied:
> a user agent may wish to acquire user consent in cases where the bounds are likely to be truly unique. 

This feels a bit at odds with the statement above about IPD fingerprinting, tbh.  I mean when you're drawing a weird shaped bounds (which the Quest also now supports) you're going to end up fingerprinted.

A possible mitigation would be a smaller random polygon inside the bounds?  But.. ick.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+The **_mandatory conditions_** for creating an XRReferenceSpace are the following:
+*   User consent must have been obtained before creating a [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) or [unbounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-unbounded) reference space. 
+*   User consent is required before creating a [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) reference space type, if the floor level will reflect the real-world location of the user’s floor.
+*   When creating a [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) or [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) reference space type, if the origin of the reference space reflects the real-world location of the user’s floor, or if the floor level is emulated and set by the user to a non-default value, then the co ...
```
>Rounding to the nearest 1cm is suggested.

Suggested?  The minimum required?  Or the exact requirement?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   User consent is required before creating a [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) reference space type, if the floor level will reflect the real-world location of the user’s floor.
+*   When creating a [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) or [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) reference space type, if the origin of the reference space reflects the real-world location of the user’s floor, or if the floor level is emulated and set by the user to a non-default value, then the co ...
+*   When creating a [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) reference space the [boundsGeometry](https://immersive-web.github.io/webxr/#dom-xrboundedreferencespace-boundsgeometry) must be rounded sufficiently to prevent fingerprinting the user’s bounds. The region represented by the rounded bounds geometry must be a subset of the original bounds. Roundin ...
+*   Creating multiple [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded), [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) or [eye-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-eye-level) reference spaces during the same session that have different [effective origins](https://immersive-web.github.io ...
```
I'm not clear on why this is true....Are you perhaps meaning the case of when multiple bounded reference spaces with different effective origins can be related to one another by successful calls to getPose()?

NellWaliczek replied:
Looks like we're continuing this discussion in https://github.com/immersive-web/webxr/pull/638#discussion_r289209306

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+    *   Enforce the same [position limiting](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#position-limiting) bounds for [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) data within all those sessions, and require that all [effective origins](https://immersive-web.github.io/webxr/#xrspace-effective-origin) fall within those  ...
+
+### Obtaining User Consent
+The implementation of user consent is a mandatory condition for some types of object creation and data access. In addition to general [considerations](https://github.com/immersive-web/privacy-and-security/blob/master/EXPLAINER.md#considerations) for obtaining user consent, the following considerations are specific to the WebXR Device API.
```
nit: The implementation of user consent -> Acquiring user consent

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+    *   Enforce the same [position limiting](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#position-limiting) bounds for [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) data within all those sessions, and require that all [effective origins](https://immersive-web.github.io/webxr/#xrspace-effective-origin) fall within those  ...
+
+### Obtaining User Consent
+The implementation of user consent is a mandatory condition for some types of object creation and data access. In addition to general [considerations](https://github.com/immersive-web/privacy-and-security/blob/master/EXPLAINER.md#considerations) for obtaining user consent, the following considerations are specific to the WebXR Device API.
```
Given the number of links back to the privacy-and-security explainer, I'm curious if there are plans to publish the document as a non-binding recommendation from the CG?  @TrevorFSmith @cwilso @AdaRoseCannon @toji Opinions on that?

NellWaliczek replied:
@TrevorFSmith @cwilso @AdaRoseCannon  I know we talked about this a few weeks ago, but I can't remember exactly the term you used @cwilso ... was it Note? Who has the action item on this?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+#### Lifetime of Consent
+The following guidelines are suggested:
+*   User consent should only be considered valid for the [browsing context](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#browsing-context) within which it was obtained.
+*   Once a specific consent is obtained for a specific [origin](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#origin-0) in a [browsing context](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#browsing-context), that [origin](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#origin-0) does not need to obtain that specific consent again during the lifetime of that [ ...
```
Just to double check... does this mean you've moved away from the per-session permission request approach you originally were talking about in the January F2F?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+The following guidelines are suggested:
+*   User consent should only be considered valid for the [browsing context](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#browsing-context) within which it was obtained.
+*   Once a specific consent is obtained for a specific [origin](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#origin-0) in a [browsing context](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#browsing-context), that [origin](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#origin-0) does not need to obtain that specific consent again during the lifetime of that [ ...
+*   The user agent must ensure that all mandatory conditions for user consent are met before creating any [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) object. As a result, the user agent may be required to ask for user consent multiple times in a [browsing context](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#browsing-context) if it is creating multiple [X ...
```
Is this on a per-permission basis or is it cumulative for a browsing context?  As in if Session 1 requests Permission A and then Session 2 requests Permission B, will new user consent be required if Session 3 requires both Permissions A and B?

NellWaliczek replied:
If we do decide to break the permissions requests down more granularly, then yes we should be explicit about this cumulative approach here.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+  </tr>
+</table>
+
+#### Camera Data and XRSession Data
```
I'm not really sure this section is accomplishing anything since the underlying tracking systems use that same camera data along with IMU data to perform the combined pose estimation.  There's no reason a webxr polyfill wouldn't do the same thing (and in fact do), so I'm not sure what this restriction gains in terms of privacy?

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+The following threats, and associated mitigations, are the basis of the above privacy requirements.
+
+#### Pose Data
+[XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose) might be used for [input sniffing](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#reading-inputs-or-input-sniffing). [Same Origin](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#same-origin-or-single-origin-only) and [Focused and Visible](https: ...
```
Shouldn't this be inclusive of all XRPose data? (same comment for all the bullets below)

johnpallett replied:
Good observation, this could use some clarification (the above recommendations do indeed cover both XRPose and XRViewerPose).

johnpallett replied:
Updated where applicable, I'm going to resolve this one. Thanks Nell!

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+Within a [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) reference space, [boundsGeometry](https://immersive-web.github.io/webxr/#dom-xrboundedreferencespace-boundsgeometry) may be used for [user fingerprinting](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#user-fingerprinting). [Rounding](https://github.com/immersive- ...
+
+#### User Consent
+Users may become [confused or fatigued](https://github.com/immersive-web/privacy-and-security/blob/master/EXPLAINER.md#permissions) if prompted excessively for user consent during a session. This is addressed by requiring user consent once, at time of session creation.
```
This feels kinda like a buried lead :p  This is maybe the biggest point in your proposal; **your stance is that consent should only ever be requested at session creation time, yes?**

johnpallett replied:
The proposed design does indeed take that stance. It's not clear that allowing user consent to be requested elsewhere in the life of a WebXR session could be supported across platforms in a predictable way. I've added more detail to #424 ([direct link to comment](https://github.com/immersive-web/webxr/issues/424#issuecomment-492857627))

johnpallett commented:
@asajeffrey thanks for the question.

> Imagine a scenario where a user navigates to a document, provides consent for entering immersive mode, and bookmarks the document, then later visits the bookmark. Do we want to require the browser to ask for consent again whenever the browsing context changed?

As written, yes that is the suggested guideline. I could see an argument that a user agent might choose to persist user consent between browsing contexts (after acquiring it the first time) if the user agent has some other strong signal that the user trusts the origin. However, there are some relevant concerns around user consent outlined [here](https://github.com/immersive-web/privacy-and-security/issues/6) that user agents should consider when making that choice.

> So user consent needs to be asked for only when the resources for the document have been reclaimed, e.g. due to memory pressure or a process being closed?

[Browsing contexts](https://www.w3.org/TR/2009/WD-html5-20090423/browsers.html#browsing-context) cover how garbage collection occurs when a browsing context is discarded, but I don't know that consent recommendations need to specifically be tied to garbage collection. Was there a specific scenario you had in mind?

asajeffrey commented:
@johnpallett the sort of thing I'm thinking about is things like bookmarks. If a user bookmarks a page and later goes back to it, they might be confused as to why sometimes it asks for all the permissions again, and sometimes doesn't. It can skip the permissions if the browser didn't get restarted (e.g. to reclaim resources) and it's being opened in the same tab (or whatever the equivalent of tab is for XR) but not otherwise?

johnpallett commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   [Visibility state](https://www.w3.org/TR/page-visibility-2/#dom-visibilitystate) of the document is "visible".
+*   [Currently focused area](https://html.spec.whatwg.org/multipage/interaction.html#currently-focused-area-of-a-top-level-browsing-context) belongs to a document whose origin is [same origin-domain](https://html.spec.whatwg.org/multipage/origin.html#same-origin-domain) with the origin of the given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).
+*   The document is [allowed to use](https://wicg.github.io/feature-policy/#allowed-to-use) all the [policy-controlled features](https://wicg.github.io/feature-policy/#policy-controlled-feature) associated with the given [sensor types](https://www.w3.org/TR/2018/CR-generic-sensor-20180320/#sensor-type) of sensors used to generate [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose) ...
+*   If the reference space is of type [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded), [floor-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) or [eye-level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-eye-level), the [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose) data must be limited to ...
```
An earlier revision did indeed break this out, but I had feedback (I believe from @toji) that different bounds seemed arbitrary and reflective of the spec design, instead of the underlying privacy threat and mitigation. In practice room scale bounds are sufficient to mitigate the privacy threat for all reference space types.

There is also a complicating factor mentioned later in the doc that makes it desirable to have a single bounds: If you create multiple reference spaces in the same session, they should all have the same maximum bounds, to prevent the creation of a series of references spaces that can be compared to indirectly get the effect of an unbounded space. That complication suggests that it'd be desirable for a single size recommendation, allowing the user agent to establish the maximum bounded region with the first reference space, instead of potentially having to 'grow' it depending upon which reference space was created in which order.

Having said that, I don't have a strong opinion... as long as the bounds mitigate the privacy threat, there is some flexibility in approach here. @toji curious to hear your thoughts.

johnpallett replied:
As one more note, a benefit of more restrictive bounds for `local/local-floor` is that it helps mitigate the threat of user fingerprinting based upon gait and height (since a more limited position makes this more difficult).

NellWaliczek replied:
>allowing the user agent to establish the maximum bounded region with the first reference space, instead of potentially having to 'grow' it depending upon which reference space was created in which order.

This is a great clarification!  Can we get this captured in the design doc?

NellWaliczek replied:
There's one other way to think about this:
If we're saying that local/local-floor < bounded-floor < unbounded, then perhaps that should be the permission being requested at session creation time.  I'd originally been hesitating about this approach because I was concerned about the implications on adding new reference spaces in the future (such as diorama), but I think we can choose to map new reference spaces to share a "safe" bounds as we design them.  For example we might say [local/local-floor/diorama] < bounded < unbounded.

My biggest concern about this is that it will encourage developers to ask for the maximum bounds "just in case".

johnpallett commented:
> @johnpallett the sort of thing I'm thinking about is things like bookmarks. If a user bookmarks a page and later goes back to it, they might be confused as to why sometimes it asks for all the permissions again, and sometimes doesn't. It can skip the permissions if the browser didn't get restarted (e.g. to reclaim resources) and it's being opened in the same tab (or whatever the equivalent of tab is for XR) but not otherwise?

That's a great observation, particularly for browser contexts that could be running for months. The scenario suggests that user agents may wish some flexibility for how they handle consent longevity, to account for these types of scenarios (in this case, expiring user consent even though the browsing context is still active).

Right now I think that flexibility exists - the design 'suggests' but does not 'require' the guidelines - but perhaps it could be worded more precisely. Any suggestions?

@toji @NellWaliczek FYI when considering spec text.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+
+The **_mandatory conditions_** for [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) creation are the following:
+*   The document is a [responsible document](https://html.spec.whatwg.org/multipage/webappapis.html#responsible-document) of a [secure context](https://w3c.github.io/webappsec-secure-contexts/#secure-contexts).
+*   [Currently focused area](https://html.spec.whatwg.org/multipage/interaction.html#currently-focused-area-of-a-top-level-browsing-context) belongs to a document whose origin is [same origin-domain](https://html.spec.whatwg.org/multipage/origin.html#same-origin-domain) with the origin of the given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).
```
I'm supportive of that approach. 👍  The interesting thing is that this would only be possible for inline sessions because they're the only kind that can be created with all tracking disabled.  And I suspect that this not being called out explicitly was a big part of my initial confusion.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+The **_mandatory conditions_** for [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) creation are the following:
+*   The document is a [responsible document](https://html.spec.whatwg.org/multipage/webappapis.html#responsible-document) of a [secure context](https://w3c.github.io/webappsec-secure-contexts/#secure-contexts).
+*   [Currently focused area](https://html.spec.whatwg.org/multipage/interaction.html#currently-focused-area-of-a-top-level-browsing-context) belongs to a document whose origin is [same origin-domain](https://html.spec.whatwg.org/multipage/origin.html#same-origin-domain) with the origin of the given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).
+*   The document is [allowed to use](https://wicg.github.io/feature-policy/#allowed-to-use) all the [policy-controlled features](https://wicg.github.io/feature-policy/#policy-controlled-feature) associated with the given [sensor types](https://www.w3.org/TR/2018/CR-generic-sensor-20180320/#sensor-type) of sensors used to generate data available in the [XRSession](https://immersive-web.github.io/w ...
```
Ah this is a case of GitHub being slightly too smart and formatting text in a way to hide the link. 

What I was asking is do we really want to be relying on the Devices and Sensors spec to be the exhaustive definition of data sources that can drive an XR system's tracking tech.  For example, the underlying accelerometer data might not actually be exposed as a Sensor on some devices.  Also cameras are a big part of XR tracking and is not exposed as a Sensor.

johnpallett replied:
We do not want the Devices and Sensors spec to be an exhaustive definition of data sources (nor is that the purpose of this section).

The purpose here is to avoid the scenario where developers use WebXR to work around the feature policies governing underlying sensor data. In the case of IMU data there is a fairly direct correlation between XRPose and IMU data, and it would be inappropriate to allow developers access to XRPose data (and thus IMU data) if feature policy prevented access to IMU data.

However, to your point the use of _camera_ data to generate XRPose falls into a different category, as it's unlikely that a developer could use XRPose to get camera data, and the camera isn't a sensor anyway.

I wonder if this section could be generalized in terms of respecting feature policies beyond sensors, but also more precise about the intent to prevent using WebXR as a 'back door' for feature policies.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   The document is a [responsible document](https://html.spec.whatwg.org/multipage/webappapis.html#responsible-document) of a [secure context](https://w3c.github.io/webappsec-secure-contexts/#secure-contexts).
+*   [Currently focused area](https://html.spec.whatwg.org/multipage/interaction.html#currently-focused-area-of-a-top-level-browsing-context) belongs to a document whose origin is [same origin-domain](https://html.spec.whatwg.org/multipage/origin.html#same-origin-domain) with the origin of the given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).
+*   The document is [allowed to use](https://wicg.github.io/feature-policy/#allowed-to-use) all the [policy-controlled features](https://wicg.github.io/feature-policy/#policy-controlled-feature) associated with the given [sensor types](https://www.w3.org/TR/2018/CR-generic-sensor-20180320/#sensor-type) of sensors used to generate data available in the [XRSession](https://immersive-web.github.io/w ...
+*   User consent is required before creating an [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) when the [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) will support the creation and use of [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) or [unbounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype ...
```
There actually was a feature added as part of the original spatial tracking explainer I wrote before TPAC 2018 that allowed a session to reject if it didn't support the requested reference space type.  (https://github.com/immersive-web/webxr/pull/409/files#diff-f02c24f867e61b8e58dc8f2c228f476cR185) There was an issue filed to determine how to better handle developers wanting to require more than one type and we ended up removing it from the designs while we sorted out the more generic approach you refer to in #424.

That aside! I think this gets to the heart of the second main pillar of the design being proposed, yes?  In addition to "should permissions be granted at session creation time?" the other big question is: are developers expected to enumerate the permissions they want or is it an all-or-nothing deal.  If the former, we'll need an API design to make that possible in a consistent way across all browsers.  In other words, what is the categorization of permissions options you're proposing? Is it just "spatial-tracking" for now?

johnpallett replied:
Agreed on all points. This doc probably isn't the right place to answer the 'all-or-nothing' session creation question unless it affects the privacy design (AFAICT it does not). Bespoke feature requests _would_ affect this privacy design and that's covered in #424.

re: "permissions" - this design intentionally avoids specifying permissions as it's not clear whether permissions are (in all cases) suitable for gaining informed user consent. If a user agent wants to use permissions that's fine, as long as the requirements here are met, but other approaches could also meet the requirements in this doc and may be preferred depending upon a user agent's privacy policies, permissions implementation, and hardware capabilities. So, the design intentionally does not specify permission categories or options.

johnpallett replied:
I've clarified in the summary that the privacy design _does not_ have an opinion on whether user consent must be 'all or nothing' or whether the user agent can disable features based upon what the user has consented to.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   [Currently focused area](https://html.spec.whatwg.org/multipage/interaction.html#currently-focused-area-of-a-top-level-browsing-context) belongs to a document whose origin is [same origin-domain](https://html.spec.whatwg.org/multipage/origin.html#same-origin-domain) with the origin of the given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).
+*   The document is [allowed to use](https://wicg.github.io/feature-policy/#allowed-to-use) all the [policy-controlled features](https://wicg.github.io/feature-policy/#policy-controlled-feature) associated with the given [sensor types](https://www.w3.org/TR/2018/CR-generic-sensor-20180320/#sensor-type) of sensors used to generate data available in the [XRSession](https://immersive-web.github.io/w ...
+*   User consent is required before creating an [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) when the [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) will support the creation and use of [bounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-bounded) or [unbounded](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype ...
+*   User consent is required before creating an [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) when the [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) will support the creation and use of a [floor level](https://immersive-web.github.io/webxr/#dom-xrreferencespacetype-floor-level) reference space type, and the floor level will reflect the real-worl ...
```
I suspect the answer to that one depends on how many UAs decide to use quantization or not.  Also, how many categories of consent exist...

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+*   User consent is required before creating an [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) on a device with configurable interpupillary distance, if subsequent calls to [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface) could return an [XRView](https://immersive-web.github.io/webxr/#xrview) for each eye.
+*   User consent is required before creating an [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) if the user agent does not otherwise mitigate sensor fingerprinting threats through data access to [XRViewerPose](https://immersive-web.github.io/webxr/#xrviewerpose-interface).
+
+If user consent is required, it must be obtained as part of the [XRRequestSession](https://immersive-web.github.io/webxr/#dom-xr-requestsession) algorithm before the new [XRSession](https://immersive-web.github.io/webxr/#xrsession-interface) object is created. If consent is not obtained, the user agent MUST [reject](https://www.w3.org/2001/tag/doc/promises-guide/#reject-promise) the promise with  ...
```
The conversation in #424 and the screenshots @blairmacintyre posted are actually super valuable and I'd love to capture some of that explanation in this design doc if possible.

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,163 @@
+
+The **_mandatory conditions_** for exposing XRViewerPose data are the following:
+*   [Visibility state](https://www.w3.org/TR/page-visibility-2/#dom-visibilitystate) of the document is "visible".
+*   [Currently focused area](https://html.spec.whatwg.org/multipage/interaction.html#currently-focused-area-of-a-top-level-browsing-context) belongs to a document whose origin is [same origin-domain](https://html.spec.whatwg.org/multipage/origin.html#same-origin-domain) with the origin of the given [active document](https://html.spec.whatwg.org/multipage/browsers.html#active-document).
```
No objection to the language.  Just wanted to make sure I was understanding it correctly :)

NellWaliczek commented:
As @johnpallett pointed out to me yesterday, I forgot that we'd already decided that WebXR disallows creating an immersive session when one already exists.  This means we wouldn't have the difficulty of new an immersive session being requested while the user was already in one.
Given that, I do have one concern (that I'm not sure how relevant it is) that if we allow granular user consent granting that developers will be encouraged to request the kitchen sink at session creation time because they won't be able to add additional user consent after the fact.

blairmacintyre commented:
Hmmm.  Sorry that I haven't commented on this yet (I will try to this week) (as some of you know, I've been heads down on something else for about a month).

I'm keen on the idea that consent for all features can be requested during session creation, of course.  For the reasons stated (fatigue).

I'm not keen on the constraint that consent must only be requested during session creation, although depending on how permissions are structured and how permission is granted/exposed it might be ok.

How do we envision permission would be conveyed back.  If I request XR, but deny a few sub-features (detailed world knowledge, or camera/sensor access), does the request succeed (since I said yes to XR)?  Will users be told I said no, or that this feature is unavailable (but not distinguish between the device abilities or my permissions)?    

Future, if a UA has the ability for users to change their mind and upgrade permissions later, how would this work?  Or are we completely disallowing this?

In our current WIP on permissions in the WebXR Viewer, the session succeeds if they say yes, even if they deny specific sub features.  We haven't really settled on if features should be "upgradable" after you say no (right now, I think the UI allows that, but it might not actually work 😄), but you can definitely turn approved features off/on over time (e.g., stop the camera for a while).

If the app can tell what you've approved or not, or if the requestSession fails if you don't approve everything, there are obvious scenarios where this is problematic.

For example, I visit an AR (XR) application, and that application has a range of modes or features, so of which I want to use, some of which I'm uncomfortable granting the required permission.  E.g., random shopping site using a "stock" Shopify e-commerce infrastructure that provides the ability for me to visualize the product in a bunch of ways, from "floating in space" to "click to pick a point then place it", to "capture the full structure and illumination of the room, with light probes and realistic illumination and occlusion".  Perhaps I never want to allow that last feature on any site.  

The site would need to request everything up front.  If it can tell I said no to something, it may refuse to run, even if I have no intention of using the part of the site/page that needs that feature (or, if I'm happy to have a less rich experience).

johnpallett commented:
> Given that, I do have one concern (that I'm not sure how relevant it is) that if we allow granular user consent granting that developers will be encouraged to request the kitchen sink at session creation time because they won't be able to add additional user consent after the fact.

Agreed. While I don't believe the current revision of the WebXR specification allows developers to request features at session creation time, this is an important consideration for #424 

One approach would be (as outlined in #424) for developers to be able to request the features they want when creating a session. This has the benefit of allowing developers to request _only_ the features they plan to use which would allow the user agent to streamline user consent (or even avoid it in some situations).

A different approach would be for user consent to shape the capabilities of the session, giving the user more control over what the session allows. For example, if a user allows floor height to be shared, but disallows location, this would suggest that it's OK to create a session that allows creation of a local-floor reference space but not an unbounded one. One consideration of this approach is that differences between the capabilities of an XRSession and the [XRDevice list of supported modes](https://immersive-web.github.io/webxr/#list-of-supported-modes) might allow developers to infer what a user consented to.

Per our in-person discussion I'll ensure the User Communication section has a complete list of feature/consent pairs.

avadacatavra commented:
@blairmacintyre beat me to the question about what happens if a user denies a subset of requested features :)

> The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.

Is this setting the scene for us to incorporate future defenses against fingerprinting into the spec? Otherwise, it seems like it's a pretty difficult ask. Do you have any references for the efficacy of quantization in preventing fingerprinting? I like the idea, but I'm also not convinced that it will work (but happy to be proven wrong). Height is a particularly difficult attribute to mask, since it's expressed in so many different ways---not just the distance from the floor, but in stride length, reach, etc. One researcher I spoke to mentioned that he had to control for height when doing some research on user fingerprinting, because it was impossible to remove from the data otherwise.

johnpallett commented:
> Is this setting the scene for us to incorporate future defenses against fingerprinting into the spec? Otherwise, it seems like it's a pretty difficult ask. Do you have any references for the efficacy of quantization in preventing fingerprinting? I like the idea, but I'm also not convinced that it will work (but happy to be proven wrong). 

[This study](https://arxiv.org/pdf/1605.08763.pdf) indicates some success with quantization of orientation data to 6-degree increments. That might be suitable for an inline session. I'm not aware of research at more granular quantization levels.

> Height is a particularly difficult attribute to mask, since it's expressed in so many different ways---not just the distance from the floor, but in stride length, reach, etc. One researcher I spoke to mentioned that he had to control for height when doing some research on user fingerprinting, because it was impossible to remove from the data otherwise.

Note the difference between data that allows user profiling (learning information about the user) vs. fingerprinting (being able to detect the same device between sessions).

Specific to height: If viewer height is based on the actual viewer height, then the data allows user profiling and requires consent. In contrast, if viewer height is emulated using a user-configured value, then the value may be unique (if there are enough digits of precision). This could enable fingerprinting but that can be mitigated through rounding.

The design generally requires user consent before enabling features or exposing data that could be used for user profiling.

avadacatavra commented:
In my experience, experiences that didn't accurately capture my height were difficult and uncomfortable to use. I guess what I'm concerned about is requiring consent to expose something that I imagine will be very common---part of me feels like height has to be part of the baseline requirements to run XR experiences to prevent objects from being placed below the floor (for example). I suppose that in AR, plane detection could help with that.

blairmacintyre commented:
@johnpallett, a lot to unpack above.  

> The rationale for why consent can only be requested during session creation is covered in more detail [here](https://github.com/immersive-web/webxr/issues/424#issuecomment-492857627).

I understand that rationale.  I disagreed with your conclusions at the time, and I still do.  A "lowest common denominator" approach is inappropriate here.  The approach you suggestion will (most likely, I think) lead to an immersive web where websites all ask for all the permissions (because they'll use frameworks that do, or because they'll cut and paste sample code), and so we'll have a situation where consumers get the choice to "give companies all your data" vs "don't use AR (or VR)".  One of the main selling points of WebXR (IMHO) is that users have the option to leverage the tech without exposing their lives to random applications and companies.

>   An explicit requirement in the privacy design is that if the user does not consent to what the session will allow, the session should not be created.

On the surface, this makes sense.  But, I thought we'd discussed this before and decided that it's a bad idea. Obvious reason:  I go to site X, which wants to let me view objects in my room.  The site really wants to use world geometry for occlusion, realistic light probes, and detect moving people and things in the camera stream to do proper occlusion. I don't care about those features, I just want to see how big the thing is in my room.  It asks for "all the data", and I say "I would like to give you basic tracking/hittesting and that's it" and the site says "Sorry, you may not use my app unless you give me all the data".   

A more user-centric view would approve the session, and simply not send in geometry or complex illumination information.  

In this scenario, the site could decide to be obnoxious anyway, but now the path of least resistance is to just use the session you've been given.  In the previous case, if the site wants to be user friendly, it must actively iterate through sessions until it gets one it can live with, potentially having the user refuse many sessions requests ("I want all the data" -> "no", "I want some of the data" -> "no", "I want a different set of some of the data" -> "no", "ok, I can live with a little data" -> "yes")

yuck.

BTW, you say below the above

> The privacy design does not address whether the user agent must fail to create a session, or whether the user agent may create a session that allows the 'partial' set of features that the user consented to, but disables the ones that the user did not consent to.

so I guess I'm being dense (and thus am super confused).  Aren't these two statements saying the opposite?

> The design requires that consent to use those features be obtained at session creation, and consent must be obtained for any features that are possible in the session being created.

Seems like a massive step backwards.  Smart phone OSs have long moved away from this approach because user's (and, thus, developers) hated it.  

In reply to @avadacatavra you said
> This study indicates some success with quantization of orientation data to 6-degree increments. That might be suitable for an inline session. I'm not aware of research at more granular quantization levels.

That study appears to be studying smartphone motions sensors used by non-AR (i.e., tilt) apps.  Seems completely different than what we're talking about here.  I don't see any evidence, for example, that quantizing data would be appropriate for either handheld AR or any sort of HMD experiences.  It's seems fairly obvious it would create a pretty substandard experience.

If there is a way to explicitly ask for `just sensors, no VR/AR` (i.e., I want to do the equivalent of what folks do with motion APIs right now to play tilt games or view 360 photos, but don't want to do AR/VR)  then quantization would be ok, I think ... heck, I'd expect quantized data from ARKit/ARCore to be better than the old accelorometer/gyro orientation data because that was bad in other ways (swimmy and non-responsive, drift, etc)

thetuvix commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,286 @@
+#### Reference Space Data
+Viewer height may be used for [user profiling](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#user-profiling). User consent is required in situations where this may be determined by real-world floor level.
+
+Viewer height may be used for [user fingerprinting](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#user-fingerprinting), particularly in situations where viewer height is emulated and set by the user. [Rounding](https://github.com/immersive-web/privacy-and-security/blob/master/POSE-AND-ENVIRONMENT.md#rounding-quantization-or-fuzzing) addresses this threa ...
```
If a device has users calibrate both a `bounded-floor` origin and a `local` origin (e.g. Oculus Quest), there would be a precise fixed transform between the two that could be fingerprinted, including both an XYZ position offset and and orientation offset.  Even if you rounded one of these (probably the `local` origin, since it's probably less noticeable to move an origin not against a floor), there may still be quite a lot of entropy to fingerprint.

thetuvix replied:
The entropy of the `local` origin pose within the `bounded-floor` space is similar to @kearwood's concern about the entropy of a resting motion controller pose within the `bounded-floor` space.

It seems like this concern will apply to any two `XRSpace`s with a static relationship.

johnpallett commented:
Thanks @blairmacintyre this captures some important questions and concerns (also @avadacatavra whose work is greatly appreciated)

Note: At the f2f @joshmarinacci had a really good idea for #424. If I understood correctly, the idea was to allow a site to indicate at session creation which features they _might_ later use. That would allow the user agent to choose whether to solicit consent at session creation, or mid-session when the features were actually used, whichever gave the best experience on the platform. I’d love to see that idea captured in #424 for consideration.

In the meantime, this design draft is written (and associated choices are made) against the current spec revision which does not include #424. Here’s some clarification and a bit more background on the design choices. Eager to hear your thoughts.

> > The rationale for why consent can only be requested during session creation is covered in more detail [here](https://github.com/immersive-web/webxr/issues/424#issuecomment-492857627).

> I understand that rationale. I disagreed with your conclusions at the time, and I still do. A "lowest common denominator" approach is inappropriate here. The approach you suggestion will (most likely, I think) lead to an immersive web where websites all ask for all the permissions (because they'll use frameworks that do, or because they'll cut and paste sample code), and so we'll have a situation where consumers get the choice to "give companies all your data" vs "don't use AR (or VR)". One of the main selling points of WebXR (IMHO) is that users have the option to leverage the tech without exposing their lives to random applications and companies.

I agree with that concern.

To further explain the rationale linked above, consider a scenario where a site asks for camera, microphone, and two other XR features mid-session; this could be a mix of XR and non-XR requests ‘on the fly’ where the user agent cannot predict and consolidate them.

On some platforms that set of feature requests might trigger an easy-to-use sequence of four trusted interface events. On other platforms this might trigger a series of four "remove-headset-approve-permissions-replace-headset" events, as the user agent takes the necessary steps to get user consent for each request. On other platforms (where an immersive session is effectively a fullscreen session) this might trigger a series of four "leave/enter fullscreen" events. On future AR platforms it’s as yet unknown what the process for getting a trusted interface will be so it's not clear what the impact might be.

That UX divergence and whether the developer can predict it is the primary concern of the rationale linked above. There could be non-obvious ramifications to timing, impact on the experience UX, user abandonment, etc. by allowing immersive experiences to be disrupted in such ways.

To your point above, this divergence *could lead developers to request all features up front* for fear that on ‘some platforms’ things won’t work well. Alternatively, they may write experiences for only some tested platforms and exclude others. Or, they may not realize there is divergence and their experience may suffer. None of these are desirable outcomes IMO.

I'd love to find a solution to user consent and trusted interfaces that avoids this type of UX divergence. I’m eager to hear more about the idea that @joshmarinacci proposed for #424.

More generally, in the design proposal there is an apparent tension between two competing principles: the desire for progressive enhancement (the ability to request features on the fly), and the desire for cross-platform consistency and predictability. This tension is reflected in two implicit design _choices_ and both are worth discussion IMO.

First, _how divergent are the user consent UX experiences and is it a problem if user consent UX is different across platforms?_ The assumption in the design doc is that the fullscreen vs. HMD experiences are sufficiently divergent to cause unpredictable UX issues for developers, and also that future platforms may introduce additional divergence.

This divergence is largely based upon a need for user consent to be presented in a trusted interface. It was pointed out at the f2f that users who get used to using a non-trusted interface for consent might then be fooled into using non-trusted interfaces for other things like navigation and password entry. This highlights why some user agents may require the use of a trusted interface for user consent, even if the spec doesn't require it.

Assuming that divergent UX is indeed a concern, then a second _choice_ is whether to prioritize progressive enhancement, or cross-platform consistency. The draft design picks consistency. The rationale behind the choice is this: While we could add support for progressive enhancement later, it seems harder to address issues caused by divergent platform experiences.

For example, extending the spec later to support progressive enhancement would be a welcome change and developers could add it to their experiences later. In contrast, if allowing a divergent user consent UX leads to problems, it’s not clear how we can undo that later. This is a design _choice_ and feedback is very welcome.

> A more user-centric view would approve the session, and simply not send in geometry or complex illumination information.

Agreed. The design as written allows for this. IMO one of the great advantages of the UX mocks you've shown is that they give the user control, but the site can't determine whether consent, or platform limitations, are preventing features from being enabled.

> > The privacy design does not address whether the user agent must fail to create a session, or whether the user agent may create a session that allows the 'partial' set of features that the user consented to, but disables the ones that the user did not consent to.
> 
> so I guess I'm being dense (and thus am super confused). Aren't these two statements saying the opposite?

They are not. Combining them here:

(1) Do not create a session that exposes data that the user didn’t consent to.
(2) To achieve (1), after determining what a user consented to, either create a session and disable features that the user _didn’t_ consent to, or fail to create the session if it’s not possible to disable the necessary features.

I added a summary to the design, hopefully this is captured more effectively.

The reason the design allows both options for (2) is to give user agents flexibility in implementation, and in how they handle user consent interfaces. Specifically, it may be impractical on some platforms to present the user with a granular set of choices, or disable certain permutations of features. Feedback welcome on this point.

> That study appears to be studying smartphone motions sensors used by non-AR (i.e., tilt) apps. Seems completely different than what we're talking about here. I don't see any evidence, for example, that quantizing data would be appropriate for either handheld AR or any sort of HMD experiences.

Agreed. I don't know of any research for what type of quantization could work for handheld or HMD experiences, or whether such quantization would mitigate fingerprinting effectively, or whether it is even necessary on some HMD platforms. In fact, on some platforms pose data may not expose a fingerprinting threat.

The wording in the design doc is intended to highlight this as a consideration for user agents, but it's not intended to mandate unnecessary quantization or user consent on platforms where it there is no actual fingerprinting threat. Feedback welcome.

>  It's seems fairly obvious it would create a pretty substandard experience.

We're going to explore options here a bit more and we'll share what we find.

> If there is a way to explicitly ask for `just sensors, no VR/AR` (i.e., I want to do the equivalent of what folks do with motion APIs right now to play tilt games or view 360 photos, but don't want to do AR/VR) then quantization would be ok, I think ... heck, I'd expect quantized data from ARKit/ARCore to be better than the old accelorometer/gyro orientation data because that was bad in other ways (swimmy and non-responsive, drift, etc)

It's also possible on some platforms that pose data might be generated using a myriad of sensors, and the algorithm to generate that pose data might sufficiently mask any lower-order hardware signal that could be fingerprinted. In those situations a user agent might reasonably determine that there is no fingerprinting risk, and thus no need for additional quantization or user consent.

Appreciate the conversation @blairmacintyre and looking forward to hearing your thoughts.

NellWaliczek commented:
Per my conversation with @ddorwin and the discussion on yesterday's call, it seemed sensible to divide up the large number of related but distinct topics covered by the discussion on this PR into separate issues to make it easier to follow.  *fingers crossed* that I've captured things correctly so far and haven't missed anything major!
Once we've gotten (most) of those issues resolved we can come back and figure out what we'd like to do with this PR.

johnpallett commented:
Note: In the design the term "suggest(ion)" is intended to propose a non-"MUST" requirement (either a "SHOULD" or "MAY" or something else like a design consideration).

NellWaliczek commented:
Hey @johnpallett,
I know we've mostly moved discussion out of this PR, so I'm sorry if posting here opens a can of worms.  I have some questions some of the content in this PR that isn't covered in any of the other issues that have been filed, so I figured it might be safe ^_^

For context, I'm trying to put the finishing touches on a PR that should encapsulate the remaining unaddressed content in this PR, and I have a few minor questions that I'd love to get clarified before I post it:

For these sentences
> * User consent must have been obtained on devices with configurable interpupillary distance before creating an XRViewerPose that will return an XRView for each eye.
> * If the device supports configurable or factory-calibrated interpupillary distance that may vary from device to device, then the XRView transform data must be rounded to prevent fingerprinting. Specific precision for rounding is at the discretion of the user agent.
> * If the XRViewerPose will include multiple XRViews for displays whose positions and orientations have been configured by the user (e.g. in a CAVE) then the XRView transform data must be rounded to prevent fingerprinting.

I'm a bit confused in trying to distinguish the difference in threat level between an n-view headset with configurable IPD and an n-view CAVE system that allows the view panel placement to be configured. Do you have any additional background information that might help out?  For example, why is user consent necessary for IPD data and not the other?

For these sentences
> * If XRViewerPose data is generated by sensors, at least one of the following conditions must be met to mitigate sensor fingerprinting threats; the user agent may choose the approach that provides the best user experience:
>   * User consent must have been obtained, or
>   * Any XRView transform data generated by sensors must be quantized to prevent sensor fingerprinting, or
>   * The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.

Two questions:

First, what's the reasoning behind these protections only needing to apply to XRViewerPose data and not XRPose data?  

Second, IIRC, the Devices and Sensors WG settled settled on requiring a clear statement of intent for access to acellerometer data.  Given that, do we still believe that the options 2 and 3 are allowable?

johnpallett commented:
@NellWaliczek these are great questions, thank you!

> I'm a bit confused in trying to distinguish the difference in threat level between an n-view headset with configurable IPD and an n-view CAVE system that allows the view panel placement to be configured... why is user consent necessary for IPD data and not the other?

In that specific case, the raw display positions in the n-view CAVE might be _fingerprintable_ (i.e. the values might be unique to that device) whereas IPD data might be used for _user profiling_ of sensitive characteristics. The fingerprinting risk can be mitigated by the user agent, but I couldn't find a way to mitigate the user profiling risk, thus user consent.

I've summarized the general principles used in this design for fingerprinting and user profiling towards the end of this comment.

For reference, [this table](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#privacy-requirements) outlines the rationale behind the different requirements.

> What's the reasoning behind these protections only needing to apply to XRViewerPose data and not XRPose data?

I believe the same protections are present for both XRPose and XRViewerPose. The [XRPose section](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#providing-xrpose-data) doesn't mention XRView, but otherwise is similar:

> >If an input to getPose includes data generated by sensors, at least one of the following conditions must be met to mitigate sensor fingerprinting threats; the user agent may choose the approach that provides the best user experience:
> > - User consent must have been obtained, or
> > - Any transform data generated by sensors must be quantized to prevent sensor fingerprinting, or
> > - The user agent must otherwise ensure that the underlying device sensors are not susceptible to sensor fingerprinting.

One thing we'd previously discussed (and which I didn't have time to address before the last f2f) was merging the requirements for XRPose and XRViewerPose to better reflect the inheritance. Please let me know if you'd still like me to do that in this PR, or if there's something else I can do to help make things clearer.

> Second, IIRC, the Devices and Sensors WG settled settled on requiring a clear statement of intent for access to acellerometer data. Given that, do we still believe that the options 2 and 3 are allowable?

Yes. IIUC the D&S WG did accept a contribution for a requestPermission() method covering the DeviceOrientation Events API. I'm told the primary motivation for this was [fingerprinting](https://sensorid.cl.cam.ac.uk/); I haven't found any other concerns that would apply to WebXR. Unless there are other concerns (feedback welcome!), options 2 and 3 should still be allowable since they mitigate the primary concern. User agents who prefer can of course still require user consent.

---

In case it's useful, here are the general principles that were used in this PR with regards to fingerprinting and user profiling:

**Fingerprinting**
- The user agent must take steps (where possible) to prevent exposing data that is unique to the device. This is true even if consent has been obtained for other reasons. For example, even if user consent is obtained to expose IPD data (for reasons of user profiling) it is still in the user's best interests that the IPD data be anonymized to prevent fingerprinting.
- Some raw sensor data might be fingerprinted (e.g. IMU data) and thus access to that raw sensor data outside of WebXR requires user consent. The user agent must either ensure WebXR data based on those sensors is anonymized sufficiently to prevent fingerprinting, otherwise user consent is required.
- Mitigating the threat of fingerprinting is a requirement, but specific mitigation details are up to the user agent who is best equipped to evaluate the actual threat on a given platform using the platform's APIs. The approaches (e.g. values for rounding thresholds) in the design are intended as suggestions, not firm requirements.

**User Profiling**
- Consent is required before exposing data that sites might use to reliably infer sensitive user characteristics (such as race, gender, or age) for some population of users. For example, [user-configured IPD data might allow a site to determine the age of some users](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.62.6181&rep=rep1&type=pdf). Note: Even if most users couldn't be profiled, if _some_ users can be reliably profiled then this requirement holds for _all_ users.
- This design prioritizes the protection of sensitive user characteristics. If there is a reasonable possibility that a reliable signal for a sensitive characteristic exists, then user consent is required.
- User profiling concerns are _not_ sufficient to require consent before exposing data that either provides a weak signal for all users, or where the information that can be inferred is not particularly sensitive. For example, the size of a user's room could (in theory) indicate a user's wealth, but it might just as easily indicate that a user is at work; this ambiguity suggests the signal is unreliable and thus user consent is not required.

klausw commented:
> "Mitigating the threat of fingerprinting is a requirement, but specific mitigation details are up to the user agent who is best equipped to evaluate the actual threat on a given platform using the platform's APIs. The approaches (e.g. values for rounding thresholds) in the design are intended as suggestions, not firm requirements."

In case it helps, here's a couple of examples to try to make this more concrete.

If a HMD has user-adjustable IPD where the in-world camera distance is configured to match, an application can get this IPD value by examining the per-eye view origins. The IPD value can potentially be used for user profiling, but if this value is provided at a very high precision it could also identify a specific device, i.e. if the underlying platform stores a measured value persistently in a user profile. In this case, rounding the application-visible IPD value to the nearest 0.1mm would help reduce the fingerprinting impact, but the user profiling issue would still remain. This type of rounding would only be recommended if it doesn't make the experience worse, and this example assumes there wouldn't be a noticeable difference in perceived world scale between a rounded value of 65.3mm vs an API-reported value of 65.324323mm.

Another example would be if a manufacturer stores per-unit calibration values in each device to compensate for small manufacturing differences, and data derived from these device specific values end up visible through the WebXR API, for example if each device ends up with unique small variations in the supplied projection matrices. This could be mitigated if the application-visible API can hide these details, for example if it uses generic projection matrices and applies per-device correction in a separate reprojection step, but this may require cooperation from the underlying platform APIs.

NellWaliczek commented:
Ok, I think I've captured these answers accurately in #746.  Please leave comments there if I've messed it up.

johnpallett commented:
#746 covers IPD well. It also addresses most of the topics in this PR. Thank you @NellWaliczek!

#748 will also extend the privacy & security explainer to clarify the sensitivity of certain profiling data and [user consent suggestions](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#user-communication)

After that, most of this PR is covered but there is still some material not addressed elsewhere. I'd suggest we do the following:
1. I'd suggest [Private Browsing modes](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#private-browsing-modes) should be part of the privacy & security explainer. I've added #750 for this.
2. [Camera Data](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#camera-data-and-xrsession-data) and privacy. @NellWaliczek you'd suggested we add this as a consideration to the design doc, do you think we should bring it forward to the explainer?
3. There is a [table showing features, corresponding threat vectors, and mitigations](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#requirements-by-feature) which shows a feature-based analysis of privacy.
4. There is a [Background](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#background) section reviewing individual threat vectors and how they are mitigated, including an [Unaddressed](https://github.com/immersive-web/webxr/blob/67b0c0992e7d82c383d619900ef110a36c3bfade/designdocs/privacy-design.md#unaddressed) section covering threats not covered.
5. The [principles for when sensitive data is/is not considered a threat](https://github.com/immersive-web/webxr/pull/638#issuecomment-506553235) is covered in the comments.

Topics 1 and 2 seem suitable for the P&S explainer.

Topics 3, 4, and 5 are background, so maybe not suitable for the explainer, but I'd rather not lose them... I've referred to them frequently during our conversations and presumably they might be useful in the future as new features get added.

One approach for 3-5 would be to rework this PR into a "WebXR Privacy Principles & Analysis" document, to cover just those topics. @NellWaliczek WDYT?

NellWaliczek commented:
Thanks for going through and ensuring nothing slipped through the cracks!

1. Sounds good!
2. Good catch. Yes, I agree it'd be a good idea to include in the privacy-security-explainer.md
3. What do you think about this being an appendix in the privacy-security-explainer.md?
4. I think most of the Background section would be covered by the existing privacy-security-explainer.md text, the proposed fixes in #748, and adding the table from bullet point 3 above.  The unaddressed stuff might be good either as it's own section or as an appendix in the privacy-security-explainer.md.  WDYT?
5. Another good catch.  Those definitions (i assume you mean the stuff at the very bottom of that comment) would probably be good right after the first or second paragraph of the privacy-security-explainer.md.  Seem reasonable?

As for this PR, maybe the best thing to do is to add a big note at the top that this was background that went into the authoring of the privacy-security-explainer.md, link to it, and just merge this document as an artifact.

johnpallett commented:
Thanks @NellWaliczek 

On (2) that SGTM, added #753 to track
On (3) that SGTM, added #754 to track
On (4) I agree, the table covers it well. I've added a note to #754 regarding the unaddressed section
On (5) that SGTM. This probably makes sense as part of #748 and I've added a comment there.

I'll add a big note at the top shortly.

johnpallett commented:
@NellWaliczek added note. Do you want to merge as an artifact? (I'm not sure whether that designates a particular process - if there's anything I need to do please LMK)

NellWaliczek commented on designdocs/privacy-design.md:
```diff
@@ -0,0 +1,289 @@
+## Note
+This document was used as background for the [Privacy and Security Explainer](https://github.com/immersive-web/webxr/blob/master/privacy-security-explainer.md) and has been superseded by that explainer.
```
Better to make the link relative so that it's not impacted by branch changes.

johnpallett replied:
Oh good call, done. Thanks @NellWaliczek
